neptune/new/sync/utils.py

```diff
@@ -0,0 +1,167 @@
+        click.echo(
+            "Exception while fetching run {}. Skipping run.".format(run_id), err=True
+        )
+        logging.exception(e)
+        return None
+
+
+_project_name_missing_message = (
+    "Project name not provided. Could not synchronize offline runs. "
+    "To synchronize offline run, specify the project name with the --project flag "
+    "or by setting the {} environment variable.".format(PROJECT_ENV_NAME)
+)
+
+
+def _project_not_found_message(project_name: QualifiedName) -> str:
+    return (
+        "Project {} not found. Could not synchronize offline runs. ".format(
+            project_name
+        )
+        + "Please ensure you specified the correct project name with the --project flag "
+        + "or with the {} environment variable, or contact Neptune for support.".format(
+            PROJECT_ENV_NAME
+        )
+    )
+
+
+def get_project(
+    project_name_flag: Optional[QualifiedName], backend: NeptuneBackend
+) -> Optional[Project]:
+    project_name = project_name_flag or QualifiedName(os.getenv(PROJECT_ENV_NAME))
+    if not project_name:
+        click.echo(textwrap.fill(_project_name_missing_message), file=sys.stderr)
+        return None
+    try:
+        return backend.get_project(project_name)
+    except ProjectNotFound:
+        click.echo(
+            textwrap.fill(_project_not_found_message(project_name)), file=sys.stderr
+        )
+        return None
+
+
+def get_qualified_name(run: ApiExperiment) -> str:
+    return "{}/{}/{}".format(run.workspace, run.project_name, run.sys_id)
+
+
+def is_valid_uuid(val: Any) -> bool:
+    try:
+        uuid.UUID(val)
+        return True
+    except ValueError:
+        return False
+
+
+def is_run_synced(run_path: Path) -> bool:
+    return all(
+        _is_execution_synced(execution_path) for execution_path in run_path.iterdir()
+    )
+
+
+def _is_execution_synced(execution_path: Path) -> bool:
+    disk_queue = DiskQueue(
+        execution_path,
+        lambda x: x.to_dict(),
+        Operation.from_dict,
+        threading.RLock(),
+    )
+    return disk_queue.is_empty()
+
+
+def iterate_experiments(
+    base_path: Path,
+) -> Iterator[Tuple[ContainerType, UniqueId, Path]]:
+    if not base_path.is_dir():
+        return
+
+    for path in base_path.iterdir():
+        match_result = re.match(CONTAINER_DIR_REGEXP, path.name)
+        container_type_raw = match_result.group(2)
+        if container_type_raw is None:
+            container_type = ContainerType.RUN
+        else:
+            container_type = ContainerType(container_type_raw)
+        unique_id = UniqueId(match_result.group(3))
+
+        yield container_type, unique_id, path
+
+
+def get_offline_runs_ids(base_path: Path) -> List[str]:
+    result = []
+    if not (base_path / OFFLINE_DIRECTORY).is_dir():
+        return []
+    for run_path in (base_path / OFFLINE_DIRECTORY).iterdir():
+        dir_name = run_path.name
+        run_id = dir_name[len("run__") :]
```

Comment:
1. what about existing offline dirs, 3e849bc9-6ab3-4c00-87ee-5a56c2a02a3c?
2. why not use fancy `CONTAINER_DIR_REGEXP` (or `split("__")``)?